### PR TITLE
Search merge comment

### PR DESCRIPTION
**Pull request by CelineChien81777**

### CODE REVIEW

File: src/main/java/data_access/InMemoryPostDataAccessObject.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package data_access;
```
Comment (JqrqmyChqn) [inline src/main/java/data_access/InMemoryPostDataAccessObject.java:1]: Seperate DAO for posts, following the single responsibility principle, good job

File: src/main/java/data_access/MongoPostDataAccessObject.java [hunk 2]
```diff
@@ -7,39 +7,95 @@
...
+        try {
+            // Use a case-insensitive regex to find matching titles
+            for (Document doc : postCollection.find(new Document("title", new Document("$regex", title).append("$options", "i")))) {
+                // System.out.println("Matching Document: " + doc.toJson()); // TODO
+                Post post = Post.fromDocument(doc);
+                if (post != null) {
+                    posts.add(post);
+                }
+            }
+        } catch (Exception e) {
```
Comment (JqrqmyChqn) [inline src/main/java/data_access/MongoPostDataAccessObject.java:65]: Good handling of errors

**Review by JqrqmyChqn — APPROVED** [review]
Not many comments to be made as this was a refactoring of files to adhere to clean architecture. Very nice work, the flow of logic is clear and abides by CA

File: src/main/java/interface_adapter/posts/PostsViewModel.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    public String getErrorMessage() {
+        return errorMessage;
+    }
+
+    public void setErrorMessage(String msg) { this.errorMessage = msg;}
+
+    public void clearPosts() {posts.clear();}
+}
```
Comment (JanosRenAI) [inline src/main/java/interface_adapter/posts/PostsViewModel.java:47]: The PostsViewModel class in the interface_adapter.posts package exemplifies a straightforward implementation of data management and state tracking, fitting well within a Model-View-ViewModel (MVVM) architectural pattern. Here's a detailed critique and suggestions for improvement, noting that it adheres to principles of Clean Architecture by clearly separating concerns:

File: src/main/java/interface_adapter/comment/CommentPresenter.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    public void prepareSuccessView(CommentOutputData data) {
+        viewModel.setComments(data.getComments());
+        viewModel.setError(null);
+    }
+
+    @Override
+    public void prepareFailView(String error) {
+        viewModel.setError(error);
+    }
+}
```
Comment (JanosRenAI) [inline src/main/java/interface_adapter/comment/CommentPresenter.java:23]: The CommentPresenter class in the interface_adapter.comment package provides a clean and straightforward example of the Presenter role in the Model-View-Presenter (MVP) architectural pattern.

File: src/main/java/interface_adapter/search/SearchController.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    private final SearchInputBoundary interactor;
+
+    public SearchController(SearchInputBoundary interactor) {
+        this.interactor = interactor;
+    }
+
+    public void performSearch(String keyword) {
+        interactor.searchPostsByTitle(keyword);
+    }
+}
```
Comment (JanosRenAI) [inline src/main/java/interface_adapter/search/SearchController.java:19]: The SearchController class in the interface_adapter.search package is designed to function as part of the Clean Architecture model, particularly aligning with the Interface Adapter layer.

**Pull request merged**